### DISCUSSION TITLE
[Router] URLSearchParams to Object

### DISCUSSION THREAD

**Discussion by bitsmyth** [vercel/next.js#31451] [Help] [answered]
I am serializing an object (with nested properties) into a `URLSearchParam`. After calling the url with the created params I am trying to deserialize the url back into an object via `router.query`. Unforntunately I can´t get it to work, because the `ParsedUrlQuery` is faulty.

The payload for  `new URLSearchParams(JSON.stringify(data));`
```
const data = {
"price": {
     "min": 0,
     "calculated": {
           "12": 0,
           "24": 0,
           "48": 0
       },
      "isFetching": false,
      "lastFetchedPriceValid": true
 }
}
```

Which results in this composed url

```
http://localhost:3000/config/checkoutB?%7B%22price%22%3A%7B%22min%22%3A0%2C%22calculated%22%3A%7B%2212%22%3A0%2C%2224%22%3A0%2C%2248%22%3A0%7D%2C%22isFetching%22%3Afalse%2C%22lastFetchedPriceValid%22%3Atrue%7D%7D=
```

But by trying to parse the "already parsed" query (query: querystring.ParsedUrlQuery) from router, it throws always errors.

```
	useEffect(() => {
		const query = router.query;
		if (isEmpty(query)) return;
		console.log(query);
                const test = JSON.parse(query);  // <- Always throwing
		// cart.localConfigurationFromURLQuery(query);
	}, [router]);
```

The output of `console.log(query)`
```
{ "{\"price\":{\"min\":0,\"calculated\":{\"12\":0,\"24\":0,\"48\":0},\"isFetching\":false,\"lastFetchedPriceValid\":true}}": "" }
```

As I understand, the parse process behind router.query is setting the whole url params as key which results in an empty value.

Any ideas how to tackle this problem?

**Comment by wwwdepot** [comment, marked as answer]
Not sure your strategy is a good practice, it's complex to parse the data. 
Your data is inside the key of it `query = { "serialized-json-data": ""}`. 

It'll be easier if data is `query = { "cart": "serialized-json-data"}`
URL: `http://localhost:3000/config/checkoutB?cart=serialized-json-data`

Even better, `http://localhost:3000/config/checkoutB?cart=cartId`, make a fetch request to backend API to get the cart data.

**Reply by bitsmyth** [reply]
Our logic doesn´t allow me to fetch by `cartId` as at the moment of creating it doesn´t exist, therefore I have to use this kind of shitty solution. I end up with this solution

```
	toURLParams() {
		let searchParams = new URLSearchParams();

		searchParams.append('items', JSON.stringify(this.items));
		searchParams.append('categories', JSON.stringify(this.categories));
		searchParams.append('services', JSON.stringify(this.services));

		return searchParams;
	}

	async loadFromURLParams(urlSearchParams: ParsedUrlQuery) {
		if (!urlSearchParams) return;

		try {
			const { categories, items, services } = urlSearchParams;

			const _items = JSON.parse(items.toString());
			const _categories = JSON.parse(categories.toString());
			const _services = JSON.parse(services.toString());

			if (_items.length > 0 && _categories.length > 0 && _services.length > 0) {
				this.items = _items;
				this.services = _services;
				this.categories = _categories;
			}
		} catch (e) {
			this.items = [];
			this.categories = [];
			this.services = [];
		}
	}
```
